portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -94,21 +127,67 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     PreparedQuery results = getAllComments();
-    
     ArrayList<Comment> comments = new ArrayList<Comment>();
+
     int numberToDisplay = getNumberToDisplay(request); 
+    String paginationInstruction = (String) request.getParameter("pageAction");
+    String searchQuery = (String) request.getParameter("search");
+    searchQuery = searchQuery.substring(1,searchQuery.length() - 1);
+    int pageToken = Integer.parseInt(
+      (String) request.getParameter("pageToken"));;
 
-    results.asIterable().forEach(comment -> {
-      comments.add(Comment.fromEntity(comment));
-    });
+    comments = getFilteredComments(results, comments, searchQuery);
+    Range<Integer> commentRange = getRangeOfCommentsToDisplay(
+      paginationInstruction, numberToDisplay, comments.size(), pageToken);
     List<Comment> commentsToDisplay = 
-      comments.subList(0, Math.min(numberToDisplay, comments.size()));
+      comments.subList(commentRange.lowerEndpoint(), 
+      commentRange.upperEndpoint());
+    int newPageToken = commentRange.lowerEndpoint();
 
-    String json = convertToJson(commentsToDisplay); 
+    String json = convertToJson(commentsToDisplay, newPageToken); 
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
+  /** Returns the range of comments that should be displayed on
+    * the site, given the user's instructions (reflecting
+    * whether to jump forward a page, jump backwards, or
+    * stay on the same page of comments).
+    * @param instruction Of the form ""next"", ""previous"", 
+        or ""none"".
+    * @param numberToDisplay Indicates the desired number of comments
+        on the page.
+    * @param totalNumberComments Indicates the number of comments currently
+        in the database.
+    * @return A range from `startIndex` to `stopIndex + 1`, where
```

Comment:
note: pageToken is missing from this java doc.

Reply:
Oops! thanks for pointing this out! fixed.